### PR TITLE
refactor: update to typescript 3.9

### PR DESCRIPTION
**Pull request by KaelWD**
Had to skip a bunch of tests that started failing for no reason, if anyone else wants to have a go at fixing them be my guest. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

### CODE REVIEW

File: packages/vuetify/src/components/VTextField/VTextField.ts [hunk 1]
```diff
@@ -120,9 +120,9 @@ export default baseMixins.extend<options>().extend({
       },
     },
     isDirty (): boolean {
-      return (this.lazyValue != null &&
-        this.lazyValue.toString().length > 0) ||
-        this.badInput
+      return this.lazyValue != null
+        ? this.lazyValue.toString().length > 0
+        : this.badInput
```
Comment (DRoet) [inline packages/vuetify/src/components/VTextField/VTextField.ts]: What do you think about:
```ts
this.lazyValue?.toString().length > 0 || this.badInput
```

Comment (KaelWD) [reply packages/vuetify/src/components/VTextField/VTextField.ts]: Tried that, if it's nully it's like `undefined > 0 || this.badInput` which is a type error

**Comment by johnleider** [comment]
Is this still being worked on?

**Comment by johnleider** [comment]
Pulse check.

**Comment by KaelWD** [comment]
Still busted, 3.8.3 seems to work though

**Comment by johnleider** [comment]
Is this still being worked on? v3 work is about to start heavily.

**Comment by KaelWD** [comment]
Tests still broken, you were trying to fix them last year but it never went everywhere: https://discordapp.com/channels/340160225338195969/519331123239583755/642416962609086474

**Comment by johnleider** [comment]
This has about 24hr to get in before v2.3 is locked.

File: packages/vuetify/babel.config.js [hunk 2]
```diff
@@ -13,6 +13,8 @@ module.exports = {
       __VUETIFY_VERSION__: vuetifyPackage.version,
       __REQUIRED_VUE__: vuetifyPackage.peerDependencies.vue,
     }],
+    '@babel/plugin-proposal-nullish-coalescing-operator',
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/babel.config.js:16]: loose mode

File: packages/vuetify/src/components/VTextField/VTextField.ts (hunk 1 above)
Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VTextField/VTextField.ts:152]: if `this.lazyValue != null` is TRUE
and `this.lazyValue.toString().length > 0` is FALSE
and `this.badInput` is TRUE

`(this.lazyValue != null && this.lazyValue.toString().length > 0) || this.badInput` === `(TRUE && FALSE) || TRUE` === `TRUE`
`this.lazyValue != null ? this.lazyValue.toString().length > 0 : this.badInput` === `TRUE ? FALSE : TRUE` === `FALSE`

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 3]
```diff
@@ -61,13 +61,13 @@ export default mixins(
     dayFormat: Function as PropType<DatePickerAllowedDatesFunction | undefined>,
     disabled: Boolean,
     events: {
-      type: [Array, Function, Object] as PropType<DatePickerEvents>,
+      type: [Array, Function, Object],
       default: () => null,
-    },
+    } as PropValidator<DatePickerEvents>,
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts:66]:
```suggestion
    } as PropValidator<DatePickerEvents | null>,
```

File: packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts [hunk 4]
```diff
@@ -45,13 +46,13 @@ export default mixins(
     disabled: Boolean,
     format: Function as PropType<DatePickerFormatter | undefined>,
     events: {
-      type: [Array, Function, Object] as PropType<DatePickerEvents>,
+      type: [Array, Function, Object],
       default: () => null,
-    },
+    } as PropValidator<DatePickerEvents>,
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VDatePicker/mixins/date-picker-table.ts:51]:
```suggestion
    } as PropValidator<DatePickerEvents | null>,
```

File: packages/vuetify/src/components/VTextField/VTextField.ts (hunk 1 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VTextField/VTextField.ts:152]: :facepalm: 

`undefined > 0` isn't an error any more so we can use `?.` here now

File: packages/vuetify/babel.config.js (hunk 2 above)
Comment (KaelWD) [reply packages/vuetify/babel.config.js]: ty

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts (hunk 3 above)
Comment (KaelWD) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts:66]: We should really be doing this pretty much everywhere, most props default to `undefined`

File: .eslintrc.js [hunk 5]
```diff
@@ -29,7 +29,13 @@ module.exports = {
     }],
     'max-statements': ['error', 24],
     'no-console': 'off',
-    'comma-dangle': ['error', 'always-multiline'],
+    'comma-dangle': ['error', {
+      arrays: 'always-multiline',
+      objects: 'always-multiline',
+      imports: 'always-multiline',
+      exports: 'always-multiline',
+      functions: 'only-multiline'
```
Comment (jacekkarczmarczyk) [inline .eslintrc.js:37]: any reason for only-multiline here?

Comment (KaelWD) [reply .eslintrc.js:37]: Didn't want to introduce unrelated code style changes in this PR. comma-dangle used to ignore functions so we have a mix of both.

File: packages/vuetify/src/components/VTreeview/__tests__/VTreeview.spec.ts [hunk 6]
```diff
@@ -117,7 +117,8 @@ describe('VTreeView.ts', () => { // eslint-disable-line max-statements
     expect(wrapper.html()).toMatchSnapshot()
   })
 
-  it('should load children when expanding', async () => {
+  // TODO: fails with TS 3.9
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/__tests__/VTreeview.spec.ts:120]: This is quite a lot of coverage loss

Comment (KaelWD) [reply packages/vuetify/src/components/VTreeview/__tests__/VTreeview.spec.ts:120]: Yep

**Comment by johnleider** [comment]
Requesting **Final** comments from @vuetifyjs/contributors @vuetifyjs/core-team

File: packages/vuetify/src/components/VTooltip/VTooltip.ts [hunk 7]
```diff
@@ -44,9 +44,6 @@ export default mixins(Colorable, Delayable, Dependent, Detachable, Menuable, Tog
       default: 'span',
     },
     transition: String,
-    zIndex: {
```
Comment (sh7dm) [inline packages/vuetify/src/components/VTooltip/VTooltip.ts:47]: Where did this go? Couldn't find this in any of the changed mixins

Comment (KaelWD) [reply packages/vuetify/src/components/VTooltip/VTooltip.ts:47]: Menuable

File: packages/vuetify/src/components/VDatePicker/VDatePicker.ts [hunk 8]
```diff
@@ -323,7 +323,7 @@ export default mixins(
         consoleWarn(`Value must be ${this.isMultiple ? 'an' : 'a'} ${expected}, got ${valueType}`, this)
       }
     },
-    isDateAllowed (value: string) {
+    isDateAllowed (value: string): boolean {
```
Comment (sh7dm) [inline packages/vuetify/src/components/VDatePicker/VDatePicker.ts:326]: This type annotation can be removed as others were for consistency, can't it?

Comment (sh7dm) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts:326]: Same for some ones lower (I think `this.$createElement` has types)

Comment (KaelWD) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts:326]: Nope, it complains without them. The return type inference seems to be getting worse with each version, you used to be able to have computeds without an explicit return type but now it's needed for methods sometimes too.

Comment (sh7dm) [reply packages/vuetify/src/components/VDatePicker/VDatePicker.ts:326]: :disappointed: okay, that's really bad. I only remember broken `this` w/o computed types. Maybe we should open an issue to Vue?

**Pull request merged**